File: windows/EMWaver/Scripting/PlotBufferStore.cs
using System;
using System.Collections.Generic;

namespace EMWaver.Scripting;

public sealed class PlotBufferStore
{
    public static PlotBufferStore Shared { get; } = new();

    public delegate byte[] Provider();

    private readonly object _lock = new();
    private readonly Dictionary<string, byte[]> _buffers = new(StringComparer.Ordinal);
    private readonly Dictionary<string, Provider> _providers = new(StringComparer.Ordinal);

    private PlotBufferStore() { }

    public void SetBuffer(string id, byte[] data)
    {
        lock (_lock)
        {
            _buffers[id] = data;
        }
    }

    public void SetProvider(string id, Provider provider)
    {
        lock (_lock)
        {
            _providers[id] = provider;
        }
    }

    public byte[] GetBytes(string id)
    {
        Provider? provider;
        byte[]? stored;
        lock (_lock)
        {
            _providers.TryGetValue(id, out provider);
            _buffers.TryGetValue(id, out stored);
        }

        if (provider != null)
        {
            try { return provider(); }
            catch { return Array.Empty<byte>(); }
        }

        return stored ?? Array.Empty<byte>();
    }
}

File: apple/EMWaverAppleCore/Sources/EMWaverScriptRuntime/PlotBufferStore.swift
/*
 * EMWaver
 * Copyright (c) 2026 Luis Marnoto
 * SPDX-License-Identifier: Apache-2.0
 */

import Foundation

public final class PlotBufferStore {
    public static let shared = PlotBufferStore()

    public typealias Provider = () -> Data

    private var buffers: [String: Data] = [:]
    private var providers: [String: Provider] = [:]
    private let lock = NSLock()

    private init() {}

    public func setBuffer(id: String, data: Data) {
        lock.lock()
        buffers[id] = data
        lock.unlock()
    }

    public func setProvider(id: String, provider: @escaping Provider) {
        lock.lock()
        providers[id] = provider
        lock.unlock()
    }

    public func getBytes(id: String) -> Data {
        lock.lock()
        let provider = providers[id]
        let stored = buffers[id]
        lock.unlock()

        if let provider {
            return provider()
        }
        return stored ?? Data()
    }
}
